Fix #207: Provide a generic sendError method that sends error output (headers + optional body)

Dagefoerde opened the pull request:
* Resolves: #207
* Target version: master 

### Summary

As a WIP prototype, declare a generic `sendError()` method sends error message headers into the active socket. Optionally, a body can be added and the socket can be closed.

This changeset illustrates my thoughts from https://github.com/CollaboraOnline/online/issues/207#issuecomment-715981040 , but for me this does not compile. `make` says `error: ‘Util’ has not been declared` :(

### TODO

- [x] Fix the `error: ‘Util’ has not been declared` error
- [x] Perhaps move the method into a more speaking namespace

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [x] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

Dagefoerde commented:
Ah, I see, apparently it is a bad idea to `#include "Socket.hpp"` from Util. Well, maybe `StreamSocket` is actually a good location for `sendError`, what do you think?

Dagefoerde commented:
Force-pushed, migrating Util::sendError to StreamSocket::sendError. For reference, the variant that I was talking about (the one that did not compile) can be inspected at https://github.com/CollaboraOnline/online/compare/b1dfeba%5E...Dagefoerde:errorheaders-207-util.

mmeeks commented:
Hey - thanks - that's great. One thing though - I'd really not like to add this to the StreamSocket class - that is a very generic base-class for binary streams. Jamming web stuff into it prolly doesn't fit well, and there is no shame in a helper function =)

Can you pull that out into some utility sendWebError() function ? or perhaps several, there were some patterns of provided default values there that might be better with a simpler method, or fn argument defaults?

Either way - a nice cleanup; good stuff.

Thanks !

Dagefoerde commented:
Hey @mmeeks, sure, I can do that. Is there a unit to add it to, or should I create a new one? Problem with the current `Util` is that it cannot include `Socket`  (perhaps because `Socket` already depends on `Util`)... and I have no experience with resolving circular dependencies properly in C++ (yet :) )

mmeeks commented:
Oh sure - so, perhaps we can whack it in net/Web.cpp / .hpp or something ? =) And yes, its quite easy to end up with some horrifically weird dependency problem - particularly when combined with the unit tests - if you add those new files (don't forget to git add them) Kara or myself can try to unwind any circular horrors =)

Thanks !

Ashod commented on net/Socket.cpp:
```diff
@@ -555,6 +555,33 @@ void StreamSocket::dumpState(std::ostream& os)
         Util::dumpHex(os, "\t\toutBuffer:\n", "\t\t", _outBuffer);
 }
 
+void StreamSocket::sendError(int errorCode,
```
In terms of interface, the shutdown logic doesn't seem to belong to a "send error" function.

I'd suggest splitting it into an explicit sendError and sendErrorAndShutdown functions.

Internally, I'd have the sendErrorAndShutdown call sendError and then call the shutdown() function.

In addition, the body and extraHeader arguments can either be used in an overloaded version of sendError, or defaulted (to std::string(), not to ""). That way, we don't need to see all the "" arguments when calling sendError.

Dagefoerde replied:
resolved in 7fa7ee5e3

Ashod commented on wsd/FileServer.cpp:
```diff
@@ -415,15 +415,11 @@ void FileServerRequestHandler::handleRequest(const HTTPRequest& request,
-                    socket->send(oss.str());
-                    socket->shutdown();
+                    std::string extraHeaders =
+                        "Expires: " + Poco::DateTimeFormatter::format(
```
We have a utility Util::getHttpTimeNow() that is preferred to Poco::DateTIme.

Dagefoerde replied:
yes, but that only works for `now`, not for `later` ;)

kendy replied:
@Dagefoerde: There is also a Util::getHttpTime() that constructs that from std::chrono::system_clock::time_point, so possible to use for "later" too :-) - but we have an Easy Hack for cleaning these up elsewhere, so not a blocker.

Ashod commented on wsd/FileServer.cpp:
```diff
@@ -494,24 +490,20 @@ void FileServerRequestHandler::sendError(int errorCode, const Poco::Net::HTTPReq
-        "User-Agent: " << WOPI_AGENT_STRING << "\r\n"
-        << extraHeader
-        << "\r\n";
+    std::string body = "";
```
std::string body; 

The default construct is already good enough, no advantage in passing an empty string.

Dagefoerde replied:
resolved in 3d9ee4292bdcb88ca66b2ca11826eccb8ee488fe

Ashod commented on wsd/FileServer.cpp:
```diff
@@ -494,24 +490,20 @@ void FileServerRequestHandler::sendError(int errorCode, const Poco::Net::HTTPReq
     if (!shortMessage.empty())
     {
+        Poco::URI requestUri(request.getURI());
+        const std::string& path = requestUri.getPath();
```
Any good reason to use & here? As path is used only once (to encode) it's better to simply pass requestUri.getPath() to the encode function.

Dagefoerde replied:
Just that that was the original code from  https://github.com/CollaboraOnline/online/blob/1ef9ad4f5688ee97468aeea32e895b11333dd8ef/wsd/FileServer.cpp#L498 . But, sure, changed in 3d9ee4292

Ashod commented on wsd/FileServer.cpp:
```diff
@@ -494,24 +490,20 @@ void FileServerRequestHandler::sendError(int errorCode, const Poco::Net::HTTPReq
-        oss << "<h1>Error: " << shortMessage << "</h1>"
-            "<p>" << longMessage << ' ' << pathSanitized << "</p>"
+        headers += "Content-Type: text/html charset=UTF-8\r\n";
+        body = "<h1>Error: " + shortMessage + "</h1>" +
```
Why can't we declare body here instead of the top of the function? Placing variables closest to the site of usage is a good practice. Also, it can be made const here.

Dagefoerde replied:
I think then we would have to duplicate the final sendError call, right? I don't think that would be preferable.

Ashod commented on net/Socket.cpp:
```diff
@@ -555,6 +555,33 @@ void StreamSocket::dumpState(std::ostream& os)
+        << extraHeader;
+    if (body.empty())
+    {
+        oss << "Content-Length: 0\r\n";
```
Why not just unconditionally set the content-length to the body size?

`oss << "Content-Length: " << body.size() << "\r\n";`

Dagefoerde replied:
Yep, changed in d398545e5

Ashod commented on net/Socket.cpp:
```diff
@@ -555,6 +555,33 @@ void StreamSocket::dumpState(std::ostream& os)
+    }
+    oss << "\r\n";
+
+    if (!body.empty())
```
Why not just unconditionally output the body? If it's empty, nothing is amended to oss.

`oss << body;`

Dagefoerde replied:
:+1:  d398545e5

Ashod commented on net/Socket.cpp:
```diff
@@ -555,6 +555,33 @@ void StreamSocket::dumpState(std::ostream& os)
+
+    std::ostringstream oss;
+    oss << "HTTP/1.1 " << errorCode << "\r\n"
+        << "Date: " << Util::getHttpTimeNow() << "\r\n"
+        << "User-Agent: " << WOPI_AGENT_STRING << "\r\n"
```
The first two << operators in each line here are unnecessary.

Dagefoerde replied:
IMO removing them from these two (now three) lines makes the whole thing look inconsistent, doesn't it?

Ashod commented on wsd/LOOLWSD.cpp:
```diff
@@ -2794,15 +2780,7 @@ class ClientRequestDispatcher : public SimpleSocketHandler
-                << errMsg;
-            socket->send(oss.str());
-            socket->shutdown();
+            socket->sendError(400, errMsg, "", true);
```
We should have named enum values for all the HTTP error codes. Having 400 hard-coded is a very error prone, hard to remember what it means, etc. Best extract them into an enum and give them standard HTTP names.

Dagefoerde replied:
Hmm, generally I'd agree. But does it make sense to define a custom enum for that? Something like this should already exist, so we could use that, right?

Ashod reviewed (changes requested):
@Dagefoerde, thanks a lot for the patch!

Looks like a nice cleanup effort :)

I imagine a network-specific utilities helper as @mmeeks suggested as the right home for this. The net/ directory seems like a natural home for it. Since this is not low-level socket logic, I'd make it explicit to be Http or Web (again, as Michael suggested already). This distinction is very important to avoid confusing the usage and dependencies.

Besides that, I've left a few suggestions to improve things a bit further. Best to do things right as we're visiting this code ;)

Incidentally, we use clang-format to format our code. You need to have .clang-format in the root of the project, and then you run it on the files/lines you changed. See https://github.com/LibreOffice/core/blob/master/.clang-format for the file itself. But this is not a show-stopper, it just makes your code more consistent with the conventions we use.

Please feel free to ask if anything I suggest isn't clear, or isn't straight-forward to do.

Happy to help as necessary, and thanks against for taking the time to do this!

Dagefoerde commented:
Thank you all for the reviews and suggestions!

I have intentionally addressed the simpler things first, hoping to simplify your review. Please have a look at the current state. If all is well, I'll refactor the `sendErro*` methods and move them to a different place.

mrkara commented:
> Thank you all for the reviews and suggestions!
> 
> I have intentionally addressed the simpler things first, hoping to simplify your review. Please have a look at the current state. If all is well, I'll refactor the `sendErro*` methods and move them to a different place.

@Dagefoerde I think we can handle the remaining nitpicks with follow-up PRs. Please do that, and let's get this PR merged before it starts gettings merge conflicts. ;)

Ashod reviewed (commented):
Looks great! Thank you. Please follow @mmeeks's note, which I echoed, which is the only real issue here in terms of design. My minor suggestions were not blockers (and indeed most were in the original code, but it's always good to improve things while we're at it, and also wanted to signal that these aren't really the preferred way, so as not to mislead).

Extracting the sendError logic might not be trivial, but it's really important, because sending HTTP error when we shouldn't can cause all sorts of unexpected bugs. Anyway, let me know if you have questions or need help.

And thanks again for a great patch.

Dagefoerde commented:
Hi Ashod, no worries, I appreciate the effort to do things right when you get the chance. I'm all for that :) Also, I am really just learning C++ so I am happy about all the advice that I can get :+1:

So, please have another look. I moved things into a `net/Http` unit and made sure that the previous occurrences of (error) header output now use that.  Please review the new unit closely - this is the first time that I have created one from scratch, so I actually don't know whether I should import more things explicitly. And, after all, I don't want to break things.

Dagefoerde commented:
Also, I have added the new unit to Makefile.am, and the code compiles locally. But here on Github, the "Android" check does not seem to be quite happy. Do I have to add it somewhere else as well?

merttumer commented:
@Dagefoerde 
Hello, For android builds, we use CMakeLists.txt for adding libraries. The file is located at android/lib/src/main/cpp/CMakeLists.txt.in
But in this case, we don't really use Http for mobile so it is best that we should add #if !MOBILEAPP for all the cases that this is used

Dagefoerde commented:
Thanks for the feedback @merttumer . This has actually made things worse, as the original code bits (that I am trying to replace) do not make that distinction yet. Specifically, these: https://github.com/CollaboraOnline/online/pull/425/files#diff-1e8a0117363b29c73167be3923db3cf6244c2b24df03d7dbe75fb62b2a7fea98L262

I suppose that part could be modified in some way, but I don't know the codebase well enough to make that change properly. Can someone help? Otherwise, I'd suggest adding `Http` to CMakeLists.txt for now (and take care of the removal in a separate issue).

mrkara commented:
Hi @Dagefoerde! :)

> Thanks for the feedback @merttumer . This has actually made things worse, as the original code bits (that I am trying to replace) do not make that distinction yet. Specifically, these: https://github.com/CollaboraOnline/online/pull/425/files#diff-1e8a0117363b29c73167be3923db3cf6244c2b24df03d7dbe75fb62b2a7fea98L262
> 
> I suppose that part could be modified in some way, but I don't know the codebase well enough to make that change properly. Can someone help? Otherwise, I'd suggest adding `Http` to CMakeLists.txt for now (and take care of the removal in a separate issue).

No worries. Getting used to a new code-base is always a bit of work. You have done a nice job so far. Here is my suggestion:

* Remove the last commit
* (Optional - can be done later:) Change the new file names to HttpHelper.cpp and HttpHelper.hpp respectively to be consistent with the namespace.
* Leave the rest to us for this time. - @merttumer or I will handle the rest. Just make sure you have checked the [*Allow edits from maintainers.*](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) box. ;)-

Thanks!

Dagefoerde commented:
Hi @mrkara  @merttumer ,

All right, thanks for all the feedback! I removed the last commit and renamed the files to HttpHelper*.

"Allow edits by maintainers" is on; have at it! 🙂

merttumer commented:
@Dagefoerde 
Hello,
So I checked your pr locally and I understood the problem. I can make the changes for you or you could do it. The problem is actually we dont use clipboardrequest stuff on mobile as well but your code is inside of one of the functions but being there is not the problem but we need to ifndef out the clipboardrequest stuff for mobile completely but that is for another work:).
So what you need to do is, remove net/Http.cpp/hpp pair for avoiding confusion, and move them to socket.hpp/cpp files where you will see sendFileAndShutdown under HttpHelper namespace, also dont forget to remove the #include headers too from the relevant files where it is used and makefile.am.
Then in ClientSession.cpp  under handleClipboardRequest function, put #if !MOBILEAPP  #endif pairs on the two lines that HttpHelper::sendErrorAndShutdown(400, socket); is used. That should be enough for now, my android build had success with this. I did not try with other branches if any error would occur, I will fix or help you fix them.

Thanks so much for the effort :)

Dagefoerde commented:
Hi,
what do the others think about moving the methods to Socket.hpp/.cpp? @mmeeks was against putting this into the socket stuff originally, as that is rather low-level stuff.
Anyway, feel free to make changes to my PR, @merttumer ! :)

merttumer commented:
Hi again :)
"what do the others think about moving the methods to Socket.hpp/.cpp? @mmeeks was against putting this into the socket stuff originally, as that is rather low-level stuff."
My mistake, i did not see that, now I updated the pr as how originally wanted

Thanks so much for the work :)

Dagefoerde commented:
Wonderful, thanks for the modifications. The checks look way better now ✅ :)

mrkara commented:
@Dagefoerde Thanks a lot for the nice work, and for keeping up with all the reviews. :1st_place_medal: :slightly_smiling_face: 

We're looking forward to your next PR, and hopefully many more after that. :rocket: 

Please feel free to follow-up with the rest of the nitpicks discussed here, or pick up a new task. You are also welcome to drop by our [IRC channel](https://webchat.freenode.net/#cool-dev) or [Telegram group](https://t.me/CollaboraOnline) to ask your questions, discuss your ideas, or just chat. :speech_balloon:
